adding mysql-to-kv templates

Sagikeren88 opened the pull request.

pavius commented on mysql-to-kv/function.yaml:
```diff
@@ -0,0 +1,53 @@
+  maxReplicas: 1
+  build:
+    commands:
+      - 'pip install requests pymysql pandas'
```
No need for quotes

pavius commented on mysql-to-kv/function.yaml:
```diff
@@ -0,0 +1,53 @@
+  description: "A simple v3io-webapi usage for PutItem using MySQL public DB"
+  handler: "main:handler"
+  env:
+    - type: value
```
1. No need for `type`
2. These values need to be passed as variables. See the values file in https://github.com/nuclio/nuclio-templates/tree/master/string-manipulator

pavius commented on mysql-to-kv/function.yaml:
```diff
@@ -0,0 +1,53 @@
+  build:
+    commands:
+      - 'pip install requests pymysql pandas'
+    baseImage: "python:3.6-jessie"
```
I would opt for `python:3.6` - it's used more often and also distributed to dark sites

pavius commented on mysql-to-kv/ingest_from_mysql.py:
```diff
@@ -0,0 +1,101 @@
+    return res
+
+
+def handler(context, event):
```
1. handle should be first function (convention)
2. you're creating a connection per invocation, use init_context (should be second function after handler - convention):

```
def init_context(context):
	
	# connect to database
	connection = ...

	# set connection as member of context
	setattr(context, 'connection', connection)
```

Then in your handler, simply use it:
```
context.connection.read_sql_query(...)
```

pavius commented on mysql-to-kv/ingest_from_mysql.py:
```diff
@@ -0,0 +1,101 @@
+    for index, row in df.iterrows():
+        body = _pack_putitem_body(index, row)
+        try:
+            res = _putitem(f'http://v3io-webapi:8081/{container}/{table}/', body, igz_user, igz_password)
```
The webapi hostname/port should also be a variable

pavius commented on mysql-to-kv/ingest_from_mysql.py:
```diff
@@ -0,0 +1,101 @@
+    PutItem request
+    """
+    headers = {'Content-Type': 'application/json', 'X-v3io-function': 'PutItem'}
+    res = requests.post(url=url, data=json.dumps(body), headers=headers, auth=HTTPBasicAuth(user, password))
```
Why not just `return requests.post(...)`?

pavius commented on mysql-to-kv/ingest_from_mysql.py:
```diff
@@ -0,0 +1,101 @@
+    dict_item = dict()
+    for k, v in row.iteritems():
+        dict_item[k] = {'N' if isinstance(v, int) else 'S': '{}'.format(v)}
+    body = {
```
do we need the body variable? 
```
return { 
    ...
}
```

pavius commented on mysql-to-kv/ingest_from_mysql.py:
```diff
@@ -0,0 +1,101 @@
+    # IGZ variables #
+    #################
+    # table
+    table = os.getenv('TABLE', 'table')
```
Once you move to variables, don't use defaults

pavius commented on mysql-to-kv/ingest_from_mysql.py:
```diff
@@ -0,0 +1,101 @@
+        try:
+            res = _putitem(f'http://v3io-webapi:8081/{container}/{table}/', body, igz_user, igz_password)
+
+        except Exception:
```
I wouldn't catch this exception. You're actually providing less information about what happened (the original exception will contain the root cause, and will result in returning 500 as well)

pavius commented on mysql-to-kv/ingest_from_mysql.py:
```diff
@@ -0,0 +1,101 @@
+                                content_type='text/plain',
+                                status_code=500)
+
+    return context.Response(body='PutItem completed successfully',
```
No need to return text with 200 (machines don't read text). In fact no need to return anything, Nuclio will handle this as 204 where applicable IIRC

Sagikeren88 commented:
@pavius  - Thanks so much for the informative CR!
fixed the code per your comments + changed the following - 
* added a value YAML
* added init_context logic
* Used v3io-frames instead of webapi
